### ISSUE TITLE
Improving restics performance on large mostly unchanging datasets

### ISSUE THREAD

**Issue by MJDSys** [restic/restic#1550]
## Output of `restic version`
Self compiled latest master

## Identified Issue
As a follow up to #1538 and #1549, on large backups with few changed files (>14M files total, probably <10000 changed) the performance is now limited by the Index.Has function (50.53% of the total cpu usage of restic during my backup).  In an effort to try and decrease backup times further, I was looking to see what could be done to further reduce the affect of this function.  There seems to be two significant sources of cpu usage:
 - Calling Mutex.Lock() (8% of the total cpu usage of restic).  This is mostly just the overhead of having a lock, and is not caused by having the lock be contested.
 - Doing the actual lookup into the map (28.04% of the total cpu usage of restic)
Note those are the total times of the entire backup, not just in the calls to Has().

Also mixed in there is the setup/teardown time for using defer.  Considering that the Go team is looking on improving restic's exact use case (calling Mutex.Unlock), it may not be worth worrying about this.

## Do you have an idea how to solve the issue?
 - For the Mutex, I think replacing it with a RWMutex may help (I need to preform more benchmarking on this to be sure).  Otherwise, I'm not sure there is a way to avoid it.  If I understand the code base correctly, finalized Indexes don't need locking as they are read only.  Is this correct?  And would it be worth trying to take advantage of this fact (with the associated complexity increase)?
 - For the map, my only thought would be to decrease the size of it.  From reading how the map lookup is done, this *might* help, but more benchmarking is needed.  Is there a reason why the index size was chosen to be what it is?
 - Alternately, restic could try to call Has() less.  Restic seems to call Has() to check if a file is present in the backup repository already.  However, I was thinking if restic sees the file in the ancestor tree, it could use that fact to avoid having to check if the repository has the file.  Does this sound reasonable?  If so, should this wait on #1494 or be done at the same time?  If this is the preferred option but you want me to wait until after #1494 before trying this option that's fine.
 - My last idea would be to reorganize the indexes so that the MasterIndex (the main direct caller of Index.Has()) would only have to search through a limited number of indexes to find the blob.  This could also help reduce memory usage by avoiding having to load all the indexes into memory at the start of restic (which consumes ~6-8G for my 14M file backup), and instead load them on demand.  But this is a large change requiring a repository format break, so I can understand if this option is currently off the table.

Alternately, I can understand if you are happy with the performance as is right now (and I am generally happy too!) and rather concentrate on other more pressing features/bugs.  I was just looking to see if there was any low hanging fruit to optimize in restic and this seems to be my largest performance bottleneck right now.  If you want to come back to this problem later, feel free to either leave it open or close this issue as suits you.

## Did restic help you or made you happy in any way?
I've been looking for a backup solution to replace Crashplan Home, most importantly be able to handle my ~7TB of data with incremental backups.  Restic is the best solution I've found so far, and I'm currently planing on using it extensively.  Especially considering how easy and fast it is to use!

**Comment by fd0** [comment]
Thank you very much for taking the time to develop all the ideas for improvement! I'll try to respond to each idea separately below.

> For the Mutex, I think replacing it with a RWMutex may help (I need to preform more benchmarking on this to be sure).

Yes please, do benchmark that. I'm not sure using an RWMutex is worth it here, but we'll see.

> If I understand the code base correctly, finalized Indexes don't need locking as they are read only. Is this correct? And would it be worth trying to take advantage of this fact (with the associated complexity increase)?

I think rather than doing that we need to tackle the problem from a different angle.

At the moment, restic reads the index files from the repo and manages them separately in memory. So if you have like five index files, there will be five `Index` structs in the `MasterIndex`. That worked quite well for a while, but may be very inefficient.

What I have in mind is at start loading all index files, but then only managing the information learned from the index files, as a single, in-memory "database". There's no reason to manage the files separately, as they are not changed (only replaced/superseded by operations like `prune` and `rebuild-index`). This would automatically mean doing only one lookup in a map instead of five separate ones.

We can then even optimize this further if it's worth it, e.g. by using an advanced data structure like a bloom filter (maybe a reverse bloom filter, but we'll see about that), so that only lookups which are likely to succeed are run against the map.

But that's rather far in the future because there are much more pressing issues (e.g. rewriting the archiver and supporting compression).

> Is there a reason why the index size was chosen to be what it is?

There's no "index size" per se: The files are uploaded during the backup process and there even may be just a single very large index (e.g. after `prune`).

> Alternately, restic could try to call Has() less. Restic seems to call Has() to check if a file is present in the backup repository already. However, I was thinking if restic sees the file in the ancestor tree, it could use that fact to avoid having to check if the repository has the file. Does this sound reasonable?

Yes, that's reasonable but makes restic more brittle in some situations: When a file is lost in the repo (e.g. due to bit rot), and the data it contains is still available and read again during the next backup, the repo repairs itself. The archiver looks up the blobs, does not find them, and they get written. We would lose that. We should also wait on #1494 before changing anything in the archiver.

> My last idea would be to reorganize the indexes so that the MasterIndex (the main direct caller of Index.Has()) would only have to search through a limited number of indexes to find the blob.

This can probably arranged when we build some better data structure for the index. Also, the memory usage is a real pain for many people, so we need to tackle that. I'll think about it ;)

Thanks again for taking the time to dump so many great ideas into our bugtracker, I appreciate it!

**Comment by MJDSys** [comment]
@fd0 Ok sounds good.  I'll take a stab at benchmarking RWLock and report back here (if I don't raise a PR if the benchmarks look useful).

It sounds like the rest will be a bigger project (as I expected) and you'd prefer to dedicated restic related energies towards adding some features (both of which sound great to me!) in the near future.  For now I'll back burner these ideas, but keep them in mind as I play with the code.  I had a couple of features I wanted to take a look into, so I'll concentrate on those first.  I'll raise separate issues for those discussions.

Thanks again for providing this awesome tool!

**Comment by alphapapa** [comment]
I think https://github.com/restic/restic/issues/1502 may be relevant here as well.

In general, at a high level, it seems like restic could benefit from being able to:

* Quickly clone a snapshot
* Quickly modify individual paths within a snapshot

I don't know how restic works internally, so maybe it already essentially does that.  But it seems to me that having the ability to do those things would make it nearly trivial to implement CrashPlan-style real-time updates, where an external process passes restic a list of paths to add to or remove from a copy of the most recent snapshot (IOW, it would replace the scan phase).

**Comment by aawsome** [comment]
#2818 should massively improve `Index.Has()` on large repositories.
